src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2016 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.branch;
+
+import com.google.common.hash.Hashing;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.model.Item;
+import hudson.model.Node;
+import hudson.model.Slave;
+import hudson.model.TopLevelItem;
+import hudson.model.listeners.ItemListener;
+import hudson.remoting.Base64;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+import jenkins.slaves.WorkspaceLocator;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Chooses manageable workspace names for branch projects.
+ * @see "JENKINS-34564"
+ */
+@Restricted(NoExternalUse.class)
+@Extension(ordinal = -100)
+public class WorkspaceLocatorImpl extends WorkspaceLocator {
+
+    private static final Logger LOGGER = Logger.getLogger(WorkspaceLocatorImpl.class.getName());
+
+    /** The most characters to allow in a workspace directory name, relative to the root. Zero to disable altogether. */
+    @SuppressWarnings("FieldMayBeFinal")
+    private static /* not final */ int PATH_MAX = Integer.getInteger(WorkspaceLocatorImpl.class.getName() + ".PATH_MAX", 80);
+
+    @Override
+    public FilePath locate(TopLevelItem item, Node node) {
+        if (PATH_MAX == 0) {
+            return null;
+        }
+        if (!(item.getParent() instanceof MultiBranchProject)) {
+            return null;
+        }
+        String minimized = minimize(item.getFullName());
+        if (node instanceof Jenkins) {
+            return ((Jenkins) node).getRootPath().child("workspace/" + minimized);
```

Comment:
In `onDeleted` you mean. Yes that would make sense, if there is ever any reason to touch that code.